test/testdata/lsp/completion/intersection.rb

```diff
@@ -0,0 +1,39 @@
+# typed: true
+
+module M
+  def foo_common_1; end
+  def foo_common_2; end
+  def foo_only_on_a; end
+end
+module N
+  def foo_common_1; end
+  def foo_common_2; end
+  def foo_only_on_b; end
+end
+
+module Nullary
+  def some_method; end
+end
+
+module Unary
+  def some_method(x); end
+end
+
+# The behavior here is the *same as* T.any. (Maybe you expected it to be the dual.)
+# See the comment in union.rb for more.
+
+def test
+  ab = T.let(T.unsafe(nil), T.all(M, N))
+  ab.foo_
```

Comment:
This seems confusing.  So if users select `foo_only_on_a` or `foo_only_on_b`, they would get a type error because it's not a valid method on this intersection type.

Reply:
...or is this needed because nilable types are implemented this way...?

Reply:
I think my reasoning is "in any case where there's a method that's available only on one component, the best hope we have to make it not confusing is to have them try it out and see the error message."

We don't get the chance to explain why some things are / are not in the completion item list. So with that in mind, we should suggest anything someone might expect to be there (even mistakenly so) and then use error messages to explain why certain things are impossible after the fact.

In particular, in this case, it's impossible to call **either** `foo_only_on_a` **or** `foo_only_on_b` because those two methods don't exist on all components of `T.all(M, N)`. So if we're only suggesting 100% valid completion items, we should leave those out.

But it's possible that they didn't write `T.all(M, N)` themselves; rather, it might have come from some sort of weird interaction with control-flow sensitive typing, which would wipe away all the methods they suspected to be there. So again: better to show them all the things from both sides, and then show them the error message which will have debugging information as to where the `T.all` originated.